proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+each transport. This achieves the goal of making the dependencies optional;
+however, it requires users to look at multiple repositories.
+
+### Configurable behavior
+
+Packages often want to cater to multiple ecosystems such as the iOS or the
+server community. While most of the technologies are shared between ecosystems
+there are often some platform specific behaviors/libraries that one might use.
+For example, on Apple's platforms `OSLog` is the canonical logging systems
+whereas the server ecosystem is mostly using `swift-log`. However, there are
+some users that prefer to use `swift-log` on Apple's platforms which means
+libraries and applications cannot use platform compiler conditionals. Package
+traits would allow to make the logging system configurable independent from the
+platform.
+
+### Replacing environment variables in Package manifests
+
+A lot of packages are using environment variables in their `Package.swift` to
+configure their package. This has various reasons such as optional dependencies
+or setting certain defines for local development. Using environment variables
+inside `Package.swift` is not officially supported and with stricter sandboxing
+rules might break. Package traits provide provide a blessed way to configure a
+package.
+
+## Proposed solution
+
+This proposal introduces a new configuration for packages called _package
+traits_. Package authors can define a set of traits in their `Package.swift`
+that their package offers which provide a way to express conditional compilation
+and optional dependencies. Furthermore, a set of default enabled traits can be
+specified.
+
+```swift
+let package = Package(
+    name: "Example",
+    traits: [
+        "Foo",
+        Trait(
+            name: "Bar",
+            enabledTraits: [ // Other traits that are enabled when this trait is being enabled
+                "Foo"
+            ]
+        )
+    ],
+    defaultTraits: [ // The set of default enabled traits
+        "Foo",
+    ],
+    /// ...
+)
+```
+
+When depending on a package a set of enabled traits can be passed and whether the
+default set of traits are disabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: ["SomeTrait"],
+            disableDefaultTraits: true
+        )
+    ),
+]
+```
+
+Another common scenario is to enable a trait of a dependency only when a trait
+of the package is enabled. The below example enables the `SomeOtherTrait` when
+the `Foo` trait of this package is enabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: [
+                "SomeTrait",
+                EnabledTrait("SomeOtherTrait", condition: .when(traits: ["Foo"])),
```

Comment:
Are traits namespaces somehow? What if a package has two dependencies each of which has a trait with the same name?

Reply:
From my reading of the pitch, traits would be per-package. ie. the `Foo` here is a trait within *this* package and `SomeOtherTrait` is a trait within `SomePackage`. Maybe that needs some clarification either way though?

Reply:
Yes they are namespaces per package. I added a section in the detailed design for this.